core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -751,7 +751,9 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
           dataAccess.rawlsBillingProjectQuery.updateBillingProjects(Seq(billingProject.copy(servicePerimeter = Option(servicePerimeterName), googleProjectNumber = Option(googleProjectNumber))))
         }
 
-        _ <- servicePerimeterService.overwriteGoogleProjectsInPerimeter(servicePerimeterName)
+        _ <- dataSource.inTransaction { dataAccess =>
+          servicePerimeterService.overwriteGoogleProjectsInPerimeter(servicePerimeterName, dataAccess)
+        }
```

Comment:
I like the idea of putting all of this into `dataSource.inTransaction`!

`retryUntilSuccessOrTimeout` is called in `overwriteGoogleProjectsInPerimeter` which may take a while. Could this result in some db locks?

Reply:
Yeah, having it reside inside of a transaction is not ideal at all. Unfortunately, `withNewWorkspaceContext` as a whole resides within a transaction, and this now needs to happen inside of that function so I'm don't know that we have a better option. I suppose we could have two versions of `overwriteGoogleProjectsInPerimeter`, one that runs inside of a transaction for use in `withNewWorkspaceContext` and one that is identical to how it used to be for use in this case. Don't know that that's really worth it tbh, but open to feedback

Reply:
I think as long as perf tests are still ok, I'm ok with this.